internal/datastore/mysql/datastore.go

```diff
@@ -196,7 +196,107 @@ func NewMySQLDatastore(uri string, options ...Option) (*Datastore, error) {
 	return store, nil
 }
 
-func newMySQLExecutor(db *sql.DB) common.ExecuteQueryFunc {
+// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+func (mds *Datastore) SnapshotReader(rev datastore.Revision) datastore.Reader {
+	createTxFunc := func(ctx context.Context) (*sql.Tx, txCleanupFunc, error) {
+		tx, err := mds.db.BeginTx(ctx, mds.readTxOptions)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return tx, tx.Rollback, nil
+	}
+
+	querySplitter := common.TupleQuerySplitter{
+		Executor:         newMySQLExecutor(mds.db),
+		UsersetBatchSize: mds.usersetBatchSize,
+	}
+
+	return &mysqlReader{
+		mds.QueryBuilder,
+		createTxFunc,
+		querySplitter,
+		buildLivingObjectFilterForRevision(rev),
+	}
+}
+
+func noCleanup() error { return nil }
+
+// TODO (@vroldanbet) dupe from postgres datastore - need to refactor
+// ReadWriteTx tarts a read/write transaction, which will be committed if no error is
+// returned and rolled back if an error is returned.
+func (mds *Datastore) ReadWriteTx(
+	ctx context.Context,
+	fn datastore.TxUserFunc,
+) (datastore.Revision, error) {
+	var err error
+	for i := uint8(0); i <= mds.maxRetries; i++ {
+		var newTxnID uint64
+		if err = BeginTxFunc(ctx, mds.db, &sql.TxOptions{Isolation: sql.LevelSerializable}, func(tx *sql.Tx) error {
+			_, err := tx.ExecContext(ctx, "SET innodb_lock_wait_timeout=1")
+			if err != nil {
+				return fmt.Errorf("unable to adjust innodb_lock_wait_timeout: %w", err)
+			}
+
+			newTxnID, err = mds.createNewTransaction(ctx, tx)
+			if err != nil {
+				return fmt.Errorf("unable to create new txn ID: %w", err)
+			}
+
+			longLivedTx := func(context.Context) (*sql.Tx, txCleanupFunc, error) {
+				return tx, noCleanup, nil
+			}
+
+			querySplitter := common.TupleQuerySplitter{
+				Executor:         newMySQLExecutor(tx),
+				UsersetBatchSize: mds.usersetBatchSize,
+			}
+
+			rwt := &mysqlReadWriteTXN{
+				&mysqlReader{
+					mds.QueryBuilder,
+					longLivedTx,
+					querySplitter,
+					currentlyLivingObjects,
+				},
+				ctx,
+				tx,
+				newTxnID,
+			}
+
+			if err := fn(ctx, rwt); err != nil {
+				return err
+			}
+
+			return nil
+		}); err != nil {
+			if errorRetryable(err) {
+				continue
+			}
+
+			return datastore.NoRevision, err
+		}
+
+		return revisionFromTransaction(newTxnID), nil
+	}
+	return datastore.NoRevision, fmt.Errorf("max retries exceeded: %w", err)
+}
+
+func errorRetryable(err error) bool {
```

Comment:
nit: I prefer `isErrorRetryable`, although I know many people dislike that naming convention